test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -961,21 +977,21 @@ Client client() {
 
         // TODO: collapse these together?
         Client nodeClient() {
-            if (closed.get()) {
-                throw new RuntimeException("already closed");
-            }
             return getOrBuildNodeClient();
         }
 
         private Client getOrBuildNodeClient() {
+            var n = nodeClient;
+            if (n != null) {
+                ensureNotClosed();
+                return n;
+            }
             synchronized (InternalTestCluster.this) {
```

Comment:
This is quite hacky, and we seem to be extending it by still synchronizing on `this` in the outer class. Perhaps we should have an explicit sync object and pass it into NodeAndClient (or access it via `InternalTestCluster.this` here).

Reply:
Hmm give me a bit of time to do a follow-up, I think we can do much better here. Now that I had time to think, I think I have a way of setting this up that doesn't need any more sync at all :) Incoming!